src/base/packer.cpp

```diff
@@ -0,0 +1,515 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * packer.cpp
+ */
+
+#ident "$Id$"
+
+#include "packer.hpp"
+#include "object_representation.h"
+
+#include <vector>
+#include <string>
+
+namespace cubpacking
+{
+#define MAX_SMALL_STRING_SIZE 255
+#define LARGE_STRING_CODE 0xff
+
+#define CHECK_RANGE(ptr, endptr, amount) \
+  do \
+    { \
+      assert ((ptr) + (amount) <= (endptr)); \
+      if ((ptr) + (amount) > (endptr)) \
+        { \
+          throw ("serialization range not properly initialized"); \
```

Comment:
As far as I know, the throw call is expensive. I think it's best to just abort or leave the assert only (and release doesn't catch it).